tests/conftest.py

```diff
@@ -101,93 +17,45 @@ def pytest_addoption(parser):
-                and "run_with_all_execution_modes_by_default" not in item.keywords
-            ):
-                item.add_marker(skip_execution_mode)
-
-
-@pytest.fixture(scope="session")
-def gcs_url_stem(request):
-    url = request.config.getoption("--bucket")
-    assert url.startswith("gs://")
-    return url
-
-
-@pytest.fixture(scope="session")
-def session_tmp_gcs_url_prefix(gcs_url_stem):
-    """
-    Sets up and tears down a temporary "directory" on GCS to be shared by all
-    of our tests.
-    """
-
-    random_hex_str = "%016x" % random.randint(0, 2 ** 64)
-    path_str = f"{getpass.getuser()}/BNTESTDATA/{random_hex_str}"
-
-    gs_url = gcs_url_stem + "/" + path_str + "/"
-    # This emits a stderr warning because the URL doesn't exist.  That's
-    # annoying but I wasn't able to find a straightforward way to avoid it.
-    assert not gsutil_path_exists(gs_url)
 
-    yield gs_url
-
-    # This will throw an exception if the URL doesn't exist at this point.
-    # Currently every test using this fixture does write some objects under this URL,
-    # *and* doesn't clean all of them up. If this changes, we may need to start
-    # handling this more gracefully.
-    gsutil_wipe_path(gs_url)
-
-
-@pytest.fixture(scope="function")
-def tmp_gcs_url_prefix(session_tmp_gcs_url_prefix, request):
-    """A temporary "directory" on GCS for a single test."""
-
-    # `gsutil` doesn't support wildcard characters which are `[]` here.
-    # This is an open issue with gsutil but till it's fixed, we are going
-    # to change the node name to not have any wildcard characters.
-    # https://github.com/GoogleCloudPlatform/gsutil/issues/290
-    node_name = request.node.name.replace("[", "_").replace("]", "")
-    return session_tmp_gcs_url_prefix + node_name + "/"
-
-
-@pytest.fixture(scope="function")
-def gcs_builder(builder, tmp_gcs_url_prefix):
-    URL_PREFIX = "gs://"
-    assert tmp_gcs_url_prefix.startswith(URL_PREFIX)
-    gcs_path = tmp_gcs_url_prefix[len(URL_PREFIX) :]
-    bucket_name, object_path = gcs_path.split("/", 1)
-
-    builder = builder.build().to_builder()
-
-    builder.set("core__persistent_cache__gcs__bucket_name", bucket_name)
-    builder.set("core__persistent_cache__gcs__object_path", object_path)
-    builder.set("core__persistent_cache__gcs__enabled", True)
-
-    return builder
+def pytest_collection_modifyitems(config, items):
+    run_slow = config.getoption("--slow")
+    skip_slow = pytest.mark.skip(reason="only runs when --slow is set")
+
+    has_gcs = config.getoption("--bucket")
+    skip_needs_gcs = pytest.mark.skip(reason="only runs when --bucket is set")
+
+    run_parallel = config.getoption("--all-execution-modes")
+
+    items_to_keep = []
+    for item in items:
+        if "slow" in item.keywords and not run_slow:
+            item.add_marker(skip_slow)
+        if "needs_gcs" in item.keywords and not has_gcs:
+            item.add_marker(skip_needs_gcs)
+        if "parallel" in item.keywords:
```

Comment:
nit: It reads a little weird since it's all `if` conditionals in the same block and only the last one is an `elif`. Either:
- make all of them `elif`s and move the continue conditionals to top, or
- add a newline before this `if`, or
- replace `elif` with `if "parallel" not in item.keywords`.

Reply:
Oh, good point. I think I'll go with the newline here.